📖 Cleanup and improve version support page (#11914)
kubernetes-sigs/cluster-api: docs/book/src/reference/versions.md

> # Cluster API version support matrix and Kubernetes version support matrix

Author: Unfortunately doctoc does not play well with book pages where we have asides with H1 tags inside them.

I run the script on the pages it covers + I'm using the Table of content feature from intellij in this page

> # Cluster API version support matrix and Kubernetes version support matrix

Reviewer: Alrighty, not a big deal

The Cluster API team actively supports the latest two minor release (N, N-1); support in this context means that we:

- Have dedicate CI signal with E2E tests, unit tests, CVE scanners etc.
- Accept bug fixes, perform golang or dependency bump, etc. 
- Periodically cut patch releases

On top of supporting the N and N-1 release, the Cluster API team also maintain CI signal for the Cluster API N-2 release 
in case we have to do an emergency patch release.  
- If there is a need for an emergency patch, e.g. to fix a critical security issue, please bring this up to maintainers
  and it will be considered on a case-by-case basis. 

All considered, each Cluster API minor release is supported for a period of roughly sixteen (12) months:

- The first eight months of this timeframe will be considered the standard support period for a minor release.
- The next four months the minor release will be considered in maintenance mode.

> - At the end of the four-month maintenance mode period, the minor release will be considered EOL (end of life) and 

Author: It is the same formula used in from https://kubernetes.io/releases/patch-releases/#support-period
For me it makes sense, release goes EOL, cherry pick will stop soon (with a little bit of flexibility to deal with what type of request we have, when we actually drop E2E test etc)

> - At the end of the four-month maintenance mode period, the minor release will be considered EOL (end of life) and 

Reviewer: Okay, got it. Sounds fine

in the `apiVersion` field of Cluster API custom resources.

An API Version is considered deprecated when a new API version is published.

API deprecation and removal follow the [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/);
Cluster API maintainers might decide to support API versions longer than what is defined in the Kubernetes policy.

| API Version | Status         | Supported Until                                                                      |
|-------------|----------------|--------------------------------------------------------------------------------------|
| v1beta1     | Supported      | no less than 9 months or 3 minor releases after a newer API version will be released |
| v1alpha4    | Not served (*) | EOL since 2023-12-05 - v1.6.0 release date                                           |
| v1alpha3    | Not served (*) | EOL since 2023-07-25 - v1.5.0 release date                                           |

(*) Cluster API stopped to serve v1alpha3 API types from the v1.5 release and v1alpha4 types starting from the v1.6 release. 

> Those types still exist in Cluster API while we work to a fix (or a workaround) for [10051](https://github.com/kubernetes-sigs/cluster-api/issues/10051). 

Reviewer 2: Still around (but fixed below to the suggestion)

```suggestion
See [11919](https://github.com/kubernetes-sigs/cluster-api/issues/11919) for details about the v1alpha3/v1alpha4 removal plan.
```

<aside class="note warning">

> <h1>Warning</h1>

Reviewer: (I'll take a look, not a blocker for this PR as this PR doesn't change anything in that regard)

Note: Removal of a deprecated APIVersion in Kubernetes can cause issues with garbage collection by the kube-controller-manager.
This means that some objects which rely on garbage collection for cleanup - e.g. MachineSets and their descendent objects, 
like Machines and InfrastructureMachines, may not be cleaned up properly if those objects were created with an APIVersion 
which is no longer served.

To avoid these issues it’s advised to ensure a restart to the kube-controller-manager is done after upgrading to a version
of Cluster API which drops support for an APIVersion - e.g. v1.5 and v1.6.

This can be accomplished with any Kubernetes control-plane rollout, including a Kubernetes version upgrade, or by manually
stopping and restarting the kube-controller-manager.

</aside>

#### Cluster API release vs contract versions

Each Cluster API contract version defines a set of rules a provider is expected to comply with in order to interact with a specific Cluster API release.
Those rules can be in the form of CustomResourceDefinition (CRD) fields and/or expected behaviors to be implemented.
See [provider contracts](../developer/providers/contracts/overview.md)

> Each Cluster API release supports only one contract version, and by convention the supported contract version matches

Author: Fixed.
FYI this is going to change when we release v1beta2 contract, because we are going to support v1beta1 for some time more (but let's worry about this later)

patch release, thus extending the support matrix for the latest supported Cluster API minor release to:
* 5 Kubernetes minor releases for the management cluster (N - N-4)
* 7 Kubernetes minor releases for the workload cluster (N - N-6)

For example, Cluster API v1.7.0 would support the following Kubernetes versions:
* v1.26.x to v1.29.x for the management cluster
* v1.24.x to v1.29.x for the workload cluster
* When Kubernetes 1.30 is released, it will be supported in v1.7.x (but not in v1.6.x)

<aside class="note warning">

<h1>Warning</h1>

Cluster API support for older Kubernetes version is not a replacement/alternative for upstream Kubernetes support policies!

Support for versions of Kubernetes which itself are out of support is limited to "Cluster API can start a Cluster with this Kubernetes version"
and "Cluster API  can upgrade to the next Kubernetes version"; it does not include any extended support to Kubernetes itself.

</aside>

<aside class="note warning">

<h1>Warning</h1>

if the changes in Cluster API required to support a new Kubernetes release are too invasive, we won't backport
it to older releases and users have to wait for the next Cluster API minor release.

> </aside>

Author: ACK, we also have "the Cluster API team will try to support it in an upcoming Cluster API
patch release" but I will keep this just in case

#### Bootstrap provider Support

The Kubeadm Control Plane requires to work with a Kubeadm Bootstrap Provider of a same version.

#### Etcd API Support

The Kubeadm Control Plane Provider talks to the API server and etcd members of every Workload Cluster whose control plane it owns.
All the Cluster API Kubeadm Control Plane Provider currently supported are using [etcd v3 API](https://etcd.io/docs/v3.2/rfc/v3api/) when communicating with etcd.

#### CoreDNS Support

Each version of the Kubeadm Control Plane can upgrade up to a max CoreDNS version.
Notably, the Max CoreDNS Version could change also with patch releases.

| KCP Version | Max CoreDNS version |
|-------------|---------------------|
| v1.5        | v1.10.1             |
| >= v1.5.1   | v1.11.1             |
| v1.6        | v1.11.1             |
| v1.7        | v1.11.1             |
| v1.8        | v1.11.3             |
| >= v1.8.9   | v1.12.0             |
| v1.9        | v1.11.3             |
| >= v1.9.4   | v1.12.0             |
| v1.10       | v1.12.0             |

See [corefile-migration](https://github.com/coredns/corefile-migration)

### Other providers

> Cluster API has a vibrant ecosystem of awesome other providers maintained by independent teams and hosted outside of

Reviewer: ```suggestion
Cluster API has a vibrant ecosystem of awesome providers maintained by independent teams and hosted outside of
```

> the the Cluster API [GitHub repository](https://github.com/kubernetes-sigs/cluster-api/).

Reviewer: ```suggestion
the Cluster API [GitHub repository](https://github.com/kubernetes-sigs/cluster-api/).
```

To understand the list of supported version of a specific provider, its own Kubernetes support matrix, supported API versions, 
supported contract version and specific skip upgrade rules, please see its documentation. Please refer to [providers list](providers.md)

> In general, if a Provider version M says it is compatible with Cluster API version N, then version MUST be compatible 

Reviewer: ```suggestion
In general, if a Provider version M says it is compatible with Cluster API version N, then it MUST be compatible 
```

with a subset of the Kubernetes versions supported by Cluster API version N.

### clusterctl

It is strongly recommended to always use the latest patch version of [clusterctl](../clusterctl/overview.md), in order to get all the fixes/latest changes.

In case of upgrades, clusterctl should be upgraded first and then used to upgrade all the other components.

## Annexes

### Kubernetes Version Support and Cluster API deployment model

The most common deployment model for Cluster API assumes Core Provider, Kubeadm Bootstrap Provider, and Kubeadm Control Plane

> and at least one infrastructure provider Provider running on the Management Cluster, all managing the lifecycle

Reviewer: ```suggestion
and at least one infrastructure provider running on the Management Cluster, all managing the lifecycle
```

> of a set of _separated_ Workload clusters.

Reviewer: ```suggestion
of a set of _separate_ Workload clusters.
```

![Management/Workload Separate Clusters](../images/management-workload-separate-clusters.png)

In this scenario, the Kubernetes version of the Management and Workload Clusters are allowed to be different.
Additionally, management Clusters and Workload Clusters can be upgraded independently and in any order.

In another deployment model for Cluster API,  the Cluster API providers are used not only to managing the

> lifecycle of _separated_ Workload clusters, but also to managing the lifecycle of the Management cluster it self.

Reviewer: ```suggestion
lifecycle of _separate_ Workload clusters, but also to manage the lifecycle of the Management cluster itself.
```

> This cluster is also referred as a "self-hosted" cluster.

Reviewer: ```suggestion
This cluster is also referred to as a "self-hosted" cluster.
```

      * [Bootstrap provider Support](#bootstrap-provider-support)
      * [Etcd API Support](#etcd-api-support)
      * [CoreDNS Support](#coredns-support)
    * [Other providers](#other-providers)
    * [clusterctl](#clusterctl)
  * [Annexes](#annexes)
    * [Kubernetes version Support and Cluster API deployment model](#kubernetes-version-support-and-cluster-api-deployment-model)
    * [Kubernetes version specific notes](#kubernetes-version-specific-notes)
<!-- TOC -->

## Version support policies

### Cluster API release support

This paragraph documents the general rules defining how we determine Cluster API supported releases.

A Cluster API release correspond to a release in the [GitHub repository](https://github.com/kubernetes-sigs/cluster-api/releases)
for this project, and the corresponding images published in the Kubernetes docker registry.

For the sake of this document, the most important artifacts included in a Cluster API release are:

- The Cluster API Core provider image
- The Kubeadm Bootstrap provider image
- The Kubeadm Control Plane provider image
- The clusterctl binary

> The Cluster API team will release a new Cluster API version approximately every four months (3 release each year).

Reviewer 2: ```suggestion
The Cluster API team will release a new Cluster API version approximately every four months (3 releases each year).
```

See [release cycle](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/release/release-cycle.md) and [release calendars](https://github.com/kubernetes-sigs/cluster-api/tree/main/docs/release/releases) for more details about Cluster API releases management.

> The Cluster API team actively supports the latest two minor release (N, N-1); support in this context means that we:

Reviewer 2: ```suggestion
The Cluster API team actively supports the latest two minor releases (N, N-1); support in this context means that we:
```

- Have CI signal with E2E tests, unit tests, CVE scans etc.
- Accept bug fixes, perform golang or dependency bumps, etc. 
- Periodically cut patch releases

> On top of supporting the N and N-1 release, the Cluster API team also maintain CI signal for the Cluster API N-2 release 

Reviewer 2: ```suggestion
On top of supporting the N and N-1 release, the Cluster API team also maintains CI signal for the Cluster API N-2 releases 
```

|-------------|----------------|----------------------------------------------------------------------------------|
| v1beta1     | Supported      | at least 9 months or 3 minor releases after a newer API version will be released |
| v1alpha4    | Not served (*) | EOL since 2023-12-05 - v1.6.0 release date                                       |
| v1alpha3    | Not served (*) | EOL since 2023-07-25 - v1.5.0 release date                                       |

(*) Cluster API stopped to serve v1alpha3 API types from the v1.5 release and v1alpha4 types starting from the v1.6 release. 
Those types still exist in Cluster API while we work to a fix (or a workaround) for [10051](https://github.com/kubernetes-sigs/cluster-api/issues/10051). 

<aside class="note warning">

<h1>Warning</h1>

Note: Removal of a deprecated APIVersion in Kubernetes can cause issues with garbage collection by the kube-controller-manager.
This means that some objects which rely on garbage collection for cleanup - e.g. MachineSets and their descendent objects, 
like Machines and InfrastructureMachines, may not be cleaned up properly if those objects were created with an APIVersion 
which is no longer served.

> To avoid these issues it’s advised to ensure a restart to the kube-controller-manager is done after upgrading to a version

Reviewer 2: ```suggestion
To avoid these issues it’s advised to ensure a restart of the kube-controller-manager is done after upgrading to a version
```

the newest API version in the same Cluster API release.

| Contract Version | Status    | Supported Until                                                               |
|------------------|-----------|-------------------------------------------------------------------------------|
| v1beta1          | Supported | After a newer API contract will be released                                   |
| v1alpha4         | EOL       | EOL since 2023-12-05 - v1.6.0 release date; removal planned for v1.13, Apr 26 |
| v1alpha3         | EOL       | EOL since 2023-07-25 - v1.5.0 release date; removal planned for v1.13, Apr 26 |

See [11919](https://github.com/kubernetes-sigs/cluster-api/issues/11919) for details about the v1alpha3/v1alpha4 removal plan.

#### Supported Cluster API - Cluster API provider version Skew

When running a Cluster API release, all the provider installed in the same management cluster MUST
implement the CustomResourceDefinition (CRD) fields and/or expected behaviors defined by the release's contract version.

As a corollary, provider's version number and provider's API version number are not required to match Cluster API versions.

<aside class="note">

> The Cluster API command line tool, `clustertcl`, will take care of ensuring all the providers are on the

Reviewer 2: ```suggestion
The Cluster API command line tool, `clusterctl`, will take care of ensuring all the providers are on the
```

Notably, version skew between various Kubernetes components also defines constraints to be observed
by Cluster API, Cluster API providers or Cluster API users when performing Kubernetes version upgrades.

In some cases, also Cluster API and/or Cluster API providers are defining additional version skew constraints. For instance:
- If you are using kubeadm as a bootstrapper, you must abide to the [kubeadm skew policy](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#version-skew-policy). 
- If you are using image builder, all the Kubernetes components on a single machine are of the same version
- If your Cluster has a defined topology, with `Cluster.spec.topology` set and referencing a ClusterClass, 
  Cluster API enforces a single Kubernetes version for all the machines in the cluster.

## Supported versions matrix by provider or component

### Core provider (`cluster-api-controller`)

The following table defines the support matrix for the Cluster API core provider.
See [Cluster API release support](#cluster-api-release-support) and [Kubernetes versions support](#kubernetes-versions-support).

> |                  | v1.8, _Maintenance Mode_ | v1.9              | v1.10             |

Reviewer 2: Should this table be (also needs version changes)

```suggestion
|                  | v1.7, _Maintenance Mode_ | v1.8              | v1.9             |
```

to reflect the correct current versions? Or is this like this because its for main and supposed to be in the 1.10 book?

supported contract version and specific skip upgrade rules, please see its documentation. Please refer to [providers list](providers.md)

In general, if a provider version M says it is compatible with Cluster API version N, then it MUST be compatible 
with a subset of the Kubernetes versions supported by Cluster API version N.

### clusterctl

It is strongly recommended to always use the latest patch version of [clusterctl](../clusterctl/overview.md), in order to get all the fixes/latest changes.

In case of upgrades, clusterctl should be upgraded first and then used to upgrade all the other components.

## Annexes

### Kubernetes version Support and Cluster API deployment model

The most common deployment model for Cluster API assumes Core provider, Kubeadm Bootstrap provider, and Kubeadm Control Plane provider
and at least one infrastructure provider running on the Management Cluster, all managing the lifecycle
of a set of _separate_ Workload clusters.

![Management/Workload Separate Clusters](../images/management-workload-separate-clusters.png)

In this scenario, the Kubernetes version of the Management and Workload Clusters are allowed to be different.

> Additionally, management Clusters and Workload Clusters can be upgraded independently and in any order.

Reviewer 2: ```suggestion
Additionally, Management Clusters and Workload Clusters can be upgraded independently and in any order.
```
